cmd/cellrangerBam/cellrangerBam.go

```diff
@@ -0,0 +1,823 @@
+		currCluster = columns[0]
+		if j > 0 && currCluster != prevCluster {
+			clusterIdx++
+		}
+		columns[2] = fmt.Sprintf("%f", parse.StringToFloat64(columns[2])/ncAvg[clusterIdx])
+		newLine = strings.Join(columns, "\t")
+		matrix[j] = newLine
+		prevCluster = currCluster
+	}
+	return matrix
+}
+
+// gfpNormalize iterates over a single-cell counts map and applies a GFP normalization factor.
+func gfpNormalize(countsMap map[string]float64, normFactor float64) {
+	var counts float64
+	for i := range countsMap {
+		counts, _ = countsMap[i]
+		countsMap[i] = counts * normFactor
+	}
+}
+
+// gfpNormFactor takes in a map of cellType-gfpReads and returns a map with cellType-gfpNormalizationValue. Normalization value is calculated by the percentage of GFP reads if all
+// cell types were equal, and dividing it by the observed percentage of GFP reads in a given cell type
+func gfpNormFactor(clusterGFP map[string]int) map[string]float64 {
+	var totalCounts, clusterCounts int
+	var actualPerc float64
+	gfpNormMap := make(map[string]float64)
+	idealPerc := 1.0 / float64(len(clusterGFP))
+	for i := range clusterGFP {
+		clusterCounts, _ = clusterGFP[i]
+		totalCounts = clusterCounts + totalCounts
+	}
+	for i := range clusterGFP {
+		actualPerc = float64(clusterGFP[i]) / float64(totalCounts)
+		gfpNormMap[i] = idealPerc / actualPerc
+	}
+	return gfpNormMap
+}
+
+// parseGfpBam is similar to the parseBam function but handles bams containing GFP reads. This function also takes in a map of cellBx-cellType and an empty map containing each cellType
+// It returns a map that contains [cellType] = gfpReads
+func parseGfpBam(gfpBam string, cellTypeMap map[string]string, clusterGFP map[string]int) map[string]int {
+	var bit int32
+	var cluster string
+	var count int
+	var gfpStats []string
+	var found bool
+
+	inChan, _ := sam.GoReadToChan(gfpBam)
+	for i := range inChan {
+		num, _, _ := sam.QueryTag(i, "xf") //xf: extra flags (cellranger flags)
+		bit = num.(int32)
+		if bit&8 == 8 {
+			cellBx, _, _ := sam.QueryTag(i, "CB")
+			cluster, found = cellTypeMap[cellBx.(string)]
+			if found {
+				count, _ = clusterGFP[cluster]
+				clusterGFP[cluster] = count + 1
+			}
+		}
+	}
+	//print out some GFP read counts / cluster
+	var gfpReads int
+	for i := range clusterGFP {
+		gfpReads, _ = clusterGFP[i]
+		gfpStats = append(gfpStats, fmt.Sprintf("found %d GFP reads in %s", gfpReads, i))
+	}
+	for _, i := range gfpStats {
+		fmt.Println(i)
+	}
+	return clusterGFP
+}
+
+// umiSaturation randomly subsets the whole bam file (10% to 100% of all reads) and calculates how many UMIs are in those subests. The output is a tab delimited text file.
+func umiSaturation(umiBxSlice []string, out *fileio.EasyWriter) {
+	var perc, randNum float64
+	var j string
+	var count int
+
+	fileio.WriteToFileHandle(out, "totalReads\tumis")
+
+	for i := 1; i <= 10; i++ {
+		perc = float64(i) / 10
+		mp := make(map[string]bool)
+		count = 0
+		for _, j = range umiBxSlice {
```

Comment:
broadly, this for loop is counting unique UMIs at different proportions of the sam file. You are correct that if the random number is larger than the proportion we continue "continue" which stops the current iteration of the for loop and goes onto the next one (aka: the next read in the sam file). If the read passes the initial test then count is the total reads and the length of the map will be the number of UMIs (if the same UMI is given to the map it will just overwrite)